R/slide.R

```diff
@@ -756,3 +776,93 @@ epi_slide_mean = function(x, col_name, ..., before, after, ref_time_values,
   return(result)
 }
 
+#' Make a complete date sequence between min(x$time_value) and max
+#' (x$time_value). Produce lists of dates before min(x$time_value) and after
+#' max(x$time_value) for padding initial and final windows to size `n`.
+#'
+#' @importFrom checkmate assert_function
+#' @noRd
+full_date_seq <- function(x, before, after, time_step) {
+  pad_early_dates <- c()
+  pad_late_dates <- c()
+
+  # If dates are one of tsibble-provided classes, can step by numeric. `tsibble`
+  # defines a step of 1 to automatically be the quantum (smallest resolvable
+  # unit) of the date class. For example, one step = 1 quarter for `yearquarter`.
+  #
+  # `tsibble` classes apparently can't be added to in different units, so even
+  # if `time_step` is provided by the user, use a unit step.
+  if (inherits(x$time_value, c("yearquarter", "yearweek", "yearmonth")) ||
+    is.numeric(x$time_value)) {
+    all_dates <- seq(min(x$time_value), max(x$time_value), by = 1L)
+
+    if (before != 0) {
+      pad_early_dates <- Start(all_dates) - before:1
+    }
+    if (after != 0) {
+      pad_late_dates <- End(all_dates) + 1:after
+    }
+  } else if (missing(time_step)) {
+    # Guess what `by` should be based on the epi_df's `time_type`.
+    ttype <- attributes(x)$metadata$time_type
+    by <- switch(ttype,
+      day = "days",
+      week = "weeks",
+      yearweek = "weeks",
+      yearmonth = "months",
+      yearquarter = "quarters",
+      year = "years",
+      NA # default value for "custom", "day-time"
+    )
+
+    if (is.na(by)) {
+      Abort(
+        c(
+          "`frollmean` requires a full window to compute a result, but
+          `time_type` associated with the epi_df was not mappable to period
+          type valid for creating a date sequence.",
+          "i" = c("The input data's `time_type` was probably `custom` or `day-time`.
+          These require also passing a `time_step` function.")
+        ),
+        class = "epiprocess__epi_slide_mean__unmappable_time_type",
+        epiprocess__time_type = ttype
+      )
+    }
+
+    # `seq` `by` arg can be any of `c("days", "weeks", "months", "quarters", "years")`.
+    all_dates <- seq(min(x$time_value), max(x$time_value), by = by)
+
+    if (before != 0) {
+      pad_early_dates <- Start(all_dates) - before:1
```

Comment:
This should be something like `time_step(before:1)` or maybe something more complicated factoring in time_value class?  So core slide impl actually grabs the expected pad.  (same for late padding)

Reply:
Good catch. I'll add this on https://github.com/cmu-delphi/epiprocess/pull/400 + a test for this problem. Should be able to use `seq(from, by, length.out)` here to avoid having to figure out the `time_step` to accompany the `by`/`time_type`.

Reply:
[Test](https://github.com/cmu-delphi/epiprocess/pull/400/commits/a429145c96c4b692ec442758630a3947ae5e9db3) and [change](https://github.com/cmu-delphi/epiprocess/pull/400/commits/4bf2a29e2583e7cafc4d6c86cca285c87c3a9f09).